Adding list of rules in ciliumspec

aanm opened the pull request:
Fixes https://github.com/cilium/cilium/issues/900
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: implemented CiliumNetworkPolicy multi rule

Extended CiliumNetworkPolicy to allow setting up multi rules on a single
specification file with the keyword `specs` (instead of `spec` on a
single policy file.

Example:
apiVersion: "cilium.io/v1"
kind: CiliumNetworkPolicy
description: "Policy to test multiple rules in a single file"
metadata:
  name: "multi-rules"
specs:
  - endpointSelector:
      matchLabels:
        app: ratings
    ingress:
    - fromEndpoints:
      - matchLabels:
          app: reviews
  - endpointSelector:
      matchLabels:
        app: details
    ingress:
    - fromEndpoints:
      - matchLabels:
          app: productpage

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: add one more example to tests multi CNP enforcement

Add a runtime test to test the Cilium Network Policy with multiple
policies enforcement on a multi node environment.

Signed-off-by: André Martins <andre@cilium.io>

aanm commented on tests/k8s/multi-node/tests/02-cnp-specs.sh:
```diff
@@ -0,0 +1,144 @@
+
+docker exec -i ${cilium_id} cilium policy get io.cilium.k8s-policy-name=multi-rules 1>/dev/null
+
+# FIXME we have 2 policies with the same labels, figure it out how to deal
```
@tgraf ^^

tgraf replied:
The default behavior on matching labels is to replace. So we will never have multiple rules with identical labels.
We could also enforce this in `Parse()` and search if multiple rules have identical labels and error out. Both is fine I think.

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -835,9 +835,6 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
 		return
 	}
 
-	// Delete an eventual existing rule with matching label
```
Why are you removing this?

aanm replied:
I thought it was a leftover, do we need it? If yes, doesn't the `opts := AddOptions{Replace: true}` do the same?

tgraf replied:
It indeed looks like a leftover. Removing this is correct.

michi-covalent reviewed (approved):
will i be able to replace existing rules with the same name with this patch, or do i need to delete the existing rules first and then create new rules with the same name?

aanm committed:
tests: adding nodeselector for bookinfo spec

Since we have a bug limiting the usage of L7 proxy on multinode this
commit adds a NodeSelector into the yaml spec files to the pods are
scheduled into the same node. This commit is a workaround and should be
removed once the L7 proxy works on a multi works on a multinode
environment.

Signed-off-by: André Martins <andre@cilium.io>
